ouroboros-consensus/src/Ouroboros/Consensus/Mempool/Impl/Pure.hs

```diff
@@ -0,0 +1,321 @@
+      needsTrace =
+        if null txIds
+        then Nothing
+        else Just $
+               TraceMempoolManuallyRemovedTxs txIds (map fst (vrInvalid vr)) (isMempoolSize is')
+  in WriteRemoveTxs is' needsTrace
+
+-- | A datatype containing the new state produced by syncing with the Ledger, a
+-- snapshot of that mempool state and, if needed, a tracing message.
+data SyncWithLedger blk = NewSyncedState (InternalState blk) (MempoolSnapshot blk TicketNo) (Maybe (TraceEventMempool blk))
+
+-- | Intepret a 'SyncWithLedger' value produced by syncing the transactions in
+--  the mempool with the current ledger state of the 'ChainDB'.
+--
+-- The transactions that exist within the mempool will be revalidated
+-- against the current ledger state. Transactions which are found to be
+-- invalid with respect to the current ledger state, will be dropped
+-- from the mempool, whereas valid transactions will remain.
+--
+-- n.b. in our current implementation, when one opens a mempool, we
+-- spawn a thread which performs this action whenever the 'ChainDB' tip
+-- point changes.
+runSyncWithLedger :: forall m blk. IOLike m
+  => StrictTVar m (InternalState blk)
+  -> SyncWithLedger blk
+  -> STM m (Maybe (TraceEventMempool blk), MempoolSnapshot blk TicketNo)
+runSyncWithLedger stateVar (NewSyncedState is msp mTrace) = do
+  writeTVar stateVar is
+  return (mTrace, msp)
+
+-- | Create a 'SyncWithLedger' value representing the values that will need to
+-- be stored for committing this synchronization with the Ledger.
+--
+-- See the documentation of 'runSyncWithLedger' for more context.
+pureSyncWithLedger :: (LedgerSupportsMempool blk, HasTxId (GenTx blk), ValidateEnvelope blk)
+                   => InternalState blk
+                   -> LedgerState blk
+                   -> LedgerConfig blk
+                   -> MempoolCapacityBytesOverride
+                   -> SyncWithLedger blk
+pureSyncWithLedger istate lstate lcfg capacityOverride =
+  let vr = validateIS istate lstate lcfg capacityOverride
+      mempoolSize = isMempoolSize istate
+      snapshot = implSnapshotFromIS istate
+      removed = map fst (vrInvalid vr)
+      mTrace = if null removed
+               then Nothing
+               else Just $
+                    TraceMempoolRemoveTxs removed mempoolSize
+  in
+    NewSyncedState (internalStateFromVR vr) snapshot mTrace
+
+-- | Get a snapshot of the mempool state that is valid with respect to
+-- the given ledger state
+pureGetSnapshotFor :: forall blk.
+                      ( LedgerSupportsMempool blk
+                      , HasTxId (GenTx blk)
+                      , ValidateEnvelope blk
+                      )
+                   => LedgerConfig blk
+                   -> ForgeLedgerState blk
+                   -> MempoolCapacityBytesOverride
+                   -> InternalState blk
+                   -> MempoolSnapshot blk TicketNo
+pureGetSnapshotFor cfg blockLedgerState capacityOverride =
+    implSnapshotFromIS
```

Comment:
Style: `implSnapshotFromIS` should be indented 2 more.

Reply:
Fixed (resolving)